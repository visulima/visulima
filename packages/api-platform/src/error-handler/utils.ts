import type { HttpError } from "http-errors";
import { StatusCodes } from "http-status-codes";
import type { ServerResponse } from "node:http";

export const setErrorHeaders = (response: ServerResponse, error: unknown): void => {
    const headers: Record<string, ReadonlyArray<string> | number | string> = (error as HttpError).headers ?? {};

    Object.keys(headers).forEach((header: string) => {
        response.setHeader(header, headers[header] as ReadonlyArray<string> | number | string);
    });
};

/**
 * Send `JSON` object
 * @param response response object
 * @param jsonBody of data
 */
export const sendJson = (response: ServerResponse, jsonBody: unknown): void => {
    // Set header to application/json
    response.setHeader("content-type", "application/json; charset=utf-8");

    response.end(JSON.stringify(jsonBody));
};

export const addStatusCodeToResponse = (response: ServerResponse, error: unknown): void => {
    // respect err.statusCode

    if ((error as HttpError).statusCode !== undefined) {
        response.statusCode = (error as HttpError).statusCode;
    }

    // respect err.status

    if ((error as HttpError).status !== undefined) {
        response.statusCode = (error as HttpError).status;
    }

    // default status code to 500
    if (response.statusCode < 400) {
        response.statusCode = StatusCodes.INTERNAL_SERVER_ERROR;
    }
};
