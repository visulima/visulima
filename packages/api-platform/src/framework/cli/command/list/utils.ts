import { readFileSync, readdirSync } from "node:fs";
import { parse } from "node:path";
import type { PackageJson } from "type-fest";

interface AppExport {
    [key: string]: any;
    app?: string | { app?: string };
}

export const ALLOWED_EXTENSIONS = [".js", ".ts", ".mjs", ".cjs"];

export const getAppWorkingDirectoryPath = (appFilePath: string): string | null => {
    let lastParsedPath = parse(appFilePath);

    // Once the following condition returns false it means we traversed the whole file system
    while (lastParsedPath.base && lastParsedPath.root !== lastParsedPath.dir) {
        const parentDirectionItems = readdirSync(lastParsedPath.dir);

        const packageJSON = parentDirectionItems.find((item) => item === "package.json");

        if (packageJSON) {
            return lastParsedPath.dir;
        }

        lastParsedPath = parse(lastParsedPath.dir);
    }

    return null;
};

export const getFrameworkName = (directory: string): "express" | "fastify" | "hapi" | "koa" | "next" | null => {
    const packageJSONFilePath = `${directory}/package.json`;

    const { dependencies } = JSON.parse(readFileSync(packageJSONFilePath).toString()) as PackageJson;

    if (dependencies?.["express"]) {
        return "express";
    }

    if (dependencies?.["koa"] && (dependencies["@koa/router"] || dependencies["koa-router"])) {
        return "koa";
    }

    if (dependencies?.["next"]) {
        return "next";
    }

    if (dependencies?.["@hapi/hapi"]) {
        return "hapi";
    }

    if (dependencies?.["fastify"]) {
        return "fastify";
    }

    return null;
};

export const getApp = (appExport: AppExport, frameworkName: "express" | "fastify" | "hapi" | "koa" | "next" | null): AppExport | string | null => {
    const isExportEmpty = Object.keys(appExport).length === 0;

    if (isExportEmpty) {
        return null;
    }

    if (frameworkName === "hapi") {
        if (typeof (appExport.app as { app?: string }).app === "string") {
            return appExport.app as { app: string };
        }

        return appExport;
    }

    return appExport.app ?? appExport;
};
