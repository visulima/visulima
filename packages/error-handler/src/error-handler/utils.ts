// Re-export all utility functions from individual files
export { addStatusCodeToResponse } from "./utils/add-status-code-to-response";
export { sendJson } from "./utils/send-json";
export { setErrorHeaders } from "./utils/set-error-headers";
