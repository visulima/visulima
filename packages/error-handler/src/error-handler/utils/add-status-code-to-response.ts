import type { ServerResponse } from "node:http";

import type { HttpError } from "http-errors";
import { StatusCodes } from "http-status-codes";

export const addStatusCodeToResponse = (response: ServerResponse, error: unknown): void => {
    const error_ = error as Partial<HttpError>;
    const candidate = Number(error_.statusCode ?? error_.status);

    if (Number.isInteger(candidate) && candidate >= 400 && candidate <= 599) {
        response.statusCode = candidate;

        return;
    }

    if (response.statusCode < 400) {
        response.statusCode = StatusCodes.INTERNAL_SERVER_ERROR;
    }
};
