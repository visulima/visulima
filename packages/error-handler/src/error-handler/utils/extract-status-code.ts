import { StatusCodes } from "http-status-codes";

export const extractStatusCode = (error: unknown, fallback: number = StatusCodes.INTERNAL_SERVER_ERROR): number => {
    if (!error) {
        return fallback;
    }

    const candidate = Number((error as { statusCode?: unknown }).statusCode ?? (error as { status?: unknown }).status);

    if (Number.isInteger(candidate) && candidate >= 400 && candidate <= 599) {
        return candidate;
    }

    return fallback;
};
