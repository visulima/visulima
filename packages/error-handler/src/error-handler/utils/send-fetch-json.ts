export const sendFetchJson = (jsonBody: unknown, status: number, contentType: string = "application/json; charset=utf-8"): Response =>
    new Response(JSON.stringify(jsonBody), {
        headers: { "content-type": contentType },
        status,
    });
