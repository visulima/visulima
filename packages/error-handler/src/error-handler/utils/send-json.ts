import type { ServerResponse } from "node:http";

/**
 * Send `JSON` object
 * @param response response object
 * @param jsonBody of data
 * @param contentType optional content-type header value
 */
export const sendJson = (response: ServerResponse, jsonBody: unknown, contentType: string = "application/json; charset=utf-8"): void => {
    // Set header to provided or default content-type
    response.setHeader("content-type", contentType);

    response.end(JSON.stringify(jsonBody));
};
