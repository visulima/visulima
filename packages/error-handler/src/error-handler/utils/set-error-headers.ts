import type { ServerResponse } from "node:http";

import type { HttpError } from "http-errors";

const setErrorHeaders = (response: ServerResponse, error: unknown): void => {
    const headers: Record<string, ReadonlyArray<string> | number | string> = (error as HttpError).headers ?? {};

    Object.keys(headers).forEach((header: string) => {
        response.setHeader(header, headers[header] as ReadonlyArray<string> | number | string);
    });
};

export default setErrorHeaders;
