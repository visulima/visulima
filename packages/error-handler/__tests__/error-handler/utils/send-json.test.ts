import { createMocks } from "node-mocks-http";
import { describe, expect, it } from "vitest";

import { sendJson } from "../../../src/error-handler/utils/send-json";

describe(sendJson, () => {
    it("should send JSON with default content type", () => {
        expect.assertions(2);

        const { res } = createMocks({
            method: "GET",
        });

        const data = { number: 42, test: "test" };

        sendJson(res, data);

        // eslint-disable-next-line no-underscore-dangle
        expect(res._getData()).toMatchSnapshot("response-data");
        expect(res.getHeader("content-type")).toBe("application/json; charset=utf-8");
    });

    it("should send JSON with custom content type", () => {
        expect.assertions(2);

        const { res } = createMocks({
            method: "GET",
        });

        const data = { message: "custom type" };

        sendJson(res, data, "application/custom+json");

        // eslint-disable-next-line no-underscore-dangle
        expect(res._getData()).toBe("{\"message\":\"custom type\"}");
        expect(res.getHeader("content-type")).toBe("application/custom+json");
    });

    it("should handle empty objects", () => {
        expect.assertions(1);

        const { res } = createMocks({
            method: "GET",
        });

        sendJson(res, {});

        // eslint-disable-next-line no-underscore-dangle
        expect(res._getData()).toBe("{}");
    });

    it("should handle arrays", () => {
        expect.assertions(1);

        const { res } = createMocks({
            method: "GET",
        });

        const data = [1, 2, 3, "test"];

        sendJson(res, data);

        // eslint-disable-next-line no-underscore-dangle
        expect(res._getData()).toBe("[1,2,3,\"test\"]");
    });

    it("should handle null and undefined values", () => {
        expect.assertions(1);

        const { res } = createMocks({
            method: "GET",
        });

        const data = { nullValue: null, undefinedValue: undefined };

        sendJson(res, data);

        // eslint-disable-next-line no-underscore-dangle
        expect(res._getData()).toBe("{\"nullValue\":null}"); // undefined values are omitted by JSON.stringify
    });

    it("should handle complex nested objects", () => {
        expect.assertions(1);

        const { res } = createMocks({
            method: "GET",
        });

        const data = {
            items: ["item1", "item2"],
            user: {
                id: 123,
                name: "John Doe",
                preferences: {
                    notifications: true,
                    theme: "dark",
                },
            },
        };

        sendJson(res, data);

        // eslint-disable-next-line no-underscore-dangle
        expect(res._getData()).toMatchSnapshot("response-data");
    });
});
