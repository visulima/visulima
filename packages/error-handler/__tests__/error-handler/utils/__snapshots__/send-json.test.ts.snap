// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendJson > should handle complex nested objects > response-data 1`] = `"{"items":["item1","item2"],"user":{"id":123,"name":"John Doe","preferences":{"notifications":true,"theme":"dark"}}}"`;

exports[`sendJson > should send JSON with default content type > response-data 1`] = `"{"number":42,"test":"test"}"`;
